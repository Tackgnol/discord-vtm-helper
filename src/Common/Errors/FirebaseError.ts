const errorName = 'FirebaseError';

class FirebaseError extends Error {
	constructor(args?: string) {
		super(args);
		this.name = errorName;
	}
}

export { FirebaseError, errorName };
