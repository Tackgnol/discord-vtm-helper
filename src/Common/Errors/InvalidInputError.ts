export class InvalidInputError extends Error {
	botMessage: string;
	constructor(botMessage: string, args?: string) {
		super(args);
		this.name = 'Invalid user input';
		this.botMessage = botMessage;
		this.message = botMessage;
	}
}

export default InvalidInputError;
