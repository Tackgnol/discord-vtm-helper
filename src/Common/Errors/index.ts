export * from './FirebaseError';
export * from './InvalidInputError';
