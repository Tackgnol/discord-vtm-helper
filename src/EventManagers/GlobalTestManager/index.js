const { isEmpty, isNil, isNaN, sortBy, join } = require('lodash');
const { RichEmbed } = require('discord.js');
const settings = require('../../../config/settings');

class GlobalTestManager {
	constructor(message, channel) {
		this.message = message;
		this.channel = channel;
	}

	performTest(testText, replyPrefix, versionOptions, shortCircut, value) {
		const messageChanel = this.channel;
		let richEmbed;
		if (messageChanel.type === 'text') {
			if (value === 0 || isNil(value)) {
				richEmbed = new RichEmbed()
					.setTitle(settings.Lines.globalTestHeader)
					.setDescription(testText)
					.setColor(settings.colors.richEmbeddedMain)
				messageChanel.send(richEmbed);
				return;
			}
			let reply;
			if (shortCircut) {
				reply = this.shortCircutedTest(versionOptions, +value);
			} else {
				reply = this.fullTest(versionOptions, +value);
			}
			richEmbed = new RichEmbed()
				.setTitle(settings.Lines.globalTestReply)
				.addField(replyPrefix, reply)
				.setColor(settings.colors.richEmbeddedMain)
			this.message.author.send(richEmbed);
		}
	}

	shortCircutedTest(versionOptions, value) {
		if (isEmpty(versionOptions) || isNaN(value)) {
			return;
		}

		const sortedOptions = sortBy(versionOptions, v => -v.minResult);
		for (let i = 0; i <= sortedOptions.length; i++) {
			const currentOption = sortedOptions[i];
			if (currentOption.minResult <= value) {
				return currentOption.resultMessage;
			}
		}
	}

	fullTest(versionOptions, value) {
		if (isEmpty(versionOptions) || isNaN(value)) {
			return;
		}
		const sortedOptions = sortBy(versionOptions, v => v.minResult);
		const messageArray = [];
		sortedOptions.forEach(o => {
			if (o.minResult <= value) {
				messageArray.push(o.resultMessage);
			}
		});
		return !isEmpty(messageArray) && join(messageArray, '\n\t');
	}
}

module.exports = GlobalTestManager;
