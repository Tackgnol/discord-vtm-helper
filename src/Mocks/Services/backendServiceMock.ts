import { IService } from '../../Services/IService';
import { IGlobalTest, INarration, INPC, IPlayer, IStat, IStatInsight, IVersionOption } from '../../Models/GameData';
import { IGame, ISessionData } from '../../Models/AppModels';
import { testNPCWithFacts, testNPCWithoutFacts } from '../SessionDataMocks/NPCMocks';
import { globalTestWithShortCircuitMock } from '../SessionDataMocks/GlobalTestMocks';
import { basicNarration } from '../SessionDataMocks/NarrationMocks';
import { playerMock } from '../SessionDataMocks/PlayerMock';
import { statInsightMock } from '../SessionDataMocks/StatInsightMocks';
import { activeSessionMock } from '../SessionDataMocks/ActiveSessionMock';
import { gameMock } from '../GameMock';

export class BackendServiceMock implements IService {
	AddFactsToNPC(playerId: string, npc: string, facts: string[], gameId: string): Promise<INPC> {
		if (!(playerId || npc || facts || gameId)) {
			return Promise.reject('Insufficient data passed');
		}
		return Promise.resolve({ ...testNPCWithFacts, facts: [...testNPCWithFacts.facts, ...facts] });
	}

	AddGlobalTest(
		name: string,
		message: string,
		shortCircuit: boolean,
		replyPrefix: string,
		globaltestoptionSet: IVersionOption[],
		channelId: string,
		gameId: string
	): Promise<IGlobalTest> {
		if (!(name || message || shortCircuit || replyPrefix || globaltestoptionSet || channelId || gameId)) {
			return Promise.reject('Insufficient data passed');
		}
		return Promise.resolve(globalTestWithShortCircuitMock);
	}

	AddNPC(name: string, callName: string, image: string, description: string, gameId: string): Promise<Omit<INPC, 'facts'>> {
		if (!(name || callName || image || description || gameId)) {
			return Promise.reject('Insufficient data passed');
		}
		return Promise.resolve(testNPCWithoutFacts);
	}

	AddNarration(name: string, image: string, narrationText: string, channelId: string, gameId: string): Promise<INarration> {
		if (!(name || image || narrationText || channelId || gameId)) {
			return Promise.reject('Insufficient data passed');
		}
		return Promise.resolve(basicNarration);
	}

	AddPlayer(name: string, id: string, statArray: IStat[], gameId: string): Promise<IPlayer> {
		if (!(name || id || statArray || gameId)) {
			return Promise.reject('Insufficient data passed');
		}
		return Promise.resolve(playerMock);
	}

	AddStatInsight(
		name: string,
		stat: string,
		value: number,
		message: string,
		channelId: string,
		gameId: string
	): Promise<IStatInsight> {
		if (!(name || stat || value || message || channelId || gameId)) {
			return Promise.reject('Insufficient data passed');
		}
		return Promise.resolve(statInsightMock);
	}

	AssignGameAdmin(playerId: string, channelId: string, gameId: string): Promise<ISessionData> {
		if (!(playerId || channelId || gameId)) {
			Promise.reject('Insufficient data passed');
		}
		return Promise.resolve(activeSessionMock);
	}

	GetEvents(channelId: string, gameId: string): Promise<ISessionData> {
		if (!(channelId || gameId)) {
			return Promise.reject('Insufficient data passed');
		}
		return Promise.resolve(activeSessionMock);
	}

	GetPlayer(playerId: string, gameId: string): Promise<IPlayer> {
		if (!(playerId || gameId)) {
			return Promise.reject('Insufficient data passed');
		}
		return Promise.resolve(playerMock);
	}

	GetUserChannels(userId: string): Promise<IGame[]> {
		if (!userId) {
			Promise.reject('Insufficient data passed');
		}
		return Promise.resolve([gameMock]);
	}
}
