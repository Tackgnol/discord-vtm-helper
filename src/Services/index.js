const LocalFileService = require('./localFileService');
const GraphqlService = require('./graphqlService');

const initializeService = (serviceType) => {
	switch (serviceType) {
	case 'offline':
		return new LocalFileService();
	case 'graphql':
		return new GraphqlService();
	default:
		throw new Error('Invalid service type in config!');
	}
};

module.exports = initializeService;
