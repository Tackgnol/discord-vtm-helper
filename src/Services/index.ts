import LocalFileService from './localFileService';
import GraphqlService from './graphqlService';
import FirebaseService from './firebaseService';
import { IService } from './IService';

const initializeService = (serviceType: string): IService => {
	switch (serviceType) {
		case 'offline':
			return new LocalFileService();
		case 'graphql':
			return new GraphqlService();
		case 'firestore':
			return new FirebaseService();
		default:
			throw new Error('Invalid service type in config!');
	}
};

export default initializeService;
