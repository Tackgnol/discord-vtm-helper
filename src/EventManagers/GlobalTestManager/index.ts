import { isEmpty, isNaN, isNil, join, sortBy } from 'lodash';
import { globalTestRichEmbedInit, reactionNumbers } from '../../Common';
import { globalTestRichEmbedResult } from '../../Common';
import { Message, MessageOptions, MessageEmbed, StringResolvable } from 'discord.js';
import { IVersionOption } from '../../Models/GameData';
import { InvalidInputError } from '../../Common/Errors';
import { IReply, ReplyType } from '../../Models/AppModels';

export class GlobalTestManager {
	performTest(
		testText: string,
		replyPrefix: string,
		versionOptions: IVersionOption[],
		shortCircuit: boolean,
		value?: string | number
	): IReply {
		let reply;

		if (value === 0 || isNil(value)) {
			return { type: ReplyType.ReactionOneTen, value: globalTestRichEmbedInit(testText) };
		}
		if (shortCircuit) {
			reply = this.shortCircuitedTest(versionOptions, +value);
		} else {
			reply = this.fullTest(versionOptions, +value);
		}
		return { type: ReplyType.Personal, value: globalTestRichEmbedResult(replyPrefix, reply) };
	}

	shortCircuitedTest(versionOptions: IVersionOption[], value: number): string | null {
		if (isEmpty(versionOptions)) {
			throw new InvalidInputError('No responses available for this test');
		}

		if (isNaN(value)) {
			throw new InvalidInputError('Given value is invalid');
		}

		const sortedOptions = sortBy(versionOptions, v => -v.minResult);
		for (let i = 0; i <= sortedOptions.length; i++) {
			const currentOption = sortedOptions[i];
			if (currentOption.minResult <= value) {
				return currentOption.resultMessage;
			}
		}

		return null;
	}

	fullTest(versionOptions: IVersionOption[], value: number): string {
		if (isEmpty(versionOptions)) {
			throw new InvalidInputError('No responses available for this test');
		}

		if (isNaN(value)) {
			throw new InvalidInputError('Given value is invalid');
		}
		const sortedOptions = sortBy(versionOptions, v => v.minResult);
		const messageArray: string[] = [];
		sortedOptions.forEach(o => {
			if (o.minResult <= value) {
				messageArray.push(o.resultMessage);
			}
		});
		return join(messageArray, '\n\t');
	}

	handleReaction = (
		shortCircuit: boolean,
		message: Message,
		replyPrefix: string,
		versionOptions: IVersionOption[],
		sendfn: (content?: StringResolvable, options?: (MessageOptions & { split: false }) | MessageEmbed) => Promise<Message>
	) => {
		const filter = (reaction: { emoji: { name: string } }, user: { id: string }) => {
			return reactionNumbers.includes(reaction.emoji.name) && user.id === message.author.id;
		};

		message.awaitReactions(filter).then(collected => {
			const reaction = collected.first();
			const value = reactionNumbers.indexOf(reaction?.emoji.name ?? '');
			let reply: string | null;
			if (shortCircuit) {
				reply = this.shortCircuitedTest(versionOptions, +value);
			} else {
				reply = this.fullTest(versionOptions, +value);
			}
			sendfn(globalTestRichEmbedResult(replyPrefix, reply));
		});
	};
}
