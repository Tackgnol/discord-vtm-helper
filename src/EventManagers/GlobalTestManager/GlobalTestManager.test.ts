import { GlobalTestManager } from './GlobalTestManager';
import {
	globalTestWithoutShortCircuitMock,
	globalTestWithShortCircuitMock,
} from '../../Mocks/SessionDataMocks/GlobalTestMocks';
import chai, { expect } from 'chai';
import { jestSnapshotPlugin } from 'mocha-chai-jest-snapshot';
import { InvalidInputError } from '../../Common/Errors/InvalidInputError';
import { BackendServiceMock } from '../../Mocks/Services/backendServiceMock';

chai.use(jestSnapshotPlugin());

describe('EventManagers >> GlobalTestManager >> GlobalTestManager', () => {
	const manager = new GlobalTestManager();
	it('can perform a short circuit test', () => {
		const { testMessage, replyPrefix, globaltestoptionSet, shortCircuit } = globalTestWithShortCircuitMock;
		const result = manager.performTest(testMessage, replyPrefix, globaltestoptionSet, shortCircuit, 5);
		expect(result).toMatchSnapshot();
	});

	it('can perform a full test', () => {
		const { testMessage, replyPrefix, globaltestoptionSet, shortCircuit } = globalTestWithoutShortCircuitMock;
		const result = manager.performTest(testMessage, replyPrefix, globaltestoptionSet, shortCircuit, 5);
		expect(result).toMatchSnapshot();
	});

	it('returns a proper value on to low of a score', () => {
		const { testMessage, replyPrefix, globaltestoptionSet, shortCircuit } = globalTestWithoutShortCircuitMock;
		const result = manager.performTest(testMessage, replyPrefix, globaltestoptionSet, shortCircuit, 0);
		expect(result).toMatchSnapshot();
	});

	it('throws an error if the value cannot be converted to a number', () => {
		const { testMessage, replyPrefix, globaltestoptionSet, shortCircuit } = globalTestWithoutShortCircuitMock;
		expect(() => {
			manager.performTest(testMessage, replyPrefix, globaltestoptionSet, shortCircuit, 'test');
		}).to.throw(InvalidInputError);
	});

	it('throws an error if there no options given', () => {
		const { testMessage, replyPrefix, shortCircuit } = globalTestWithoutShortCircuitMock;
		expect(() => {
			manager.performTest(testMessage, replyPrefix, [], shortCircuit, 3);
		}).to.throw(InvalidInputError);
	});
});
