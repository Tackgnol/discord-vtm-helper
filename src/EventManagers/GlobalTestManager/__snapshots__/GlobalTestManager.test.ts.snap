// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EventManagers >> GlobalTestManager >> GlobalTestManager can perform a full test 1`] = `
Object {
  "type": 0,
  "value": Object {
    "author": null,
    "color": 9044739,
    "description": null,
    "fields": Array [
      Object {
        "inline": false,
        "name": "testPrefix",
        "value": "result for 1 result for 5",
      },
    ],
    "footer": null,
    "image": null,
    "thumbnail": null,
    "timestamp": null,
    "title": "Trust thine instincts, however deceitful they may be",
    "type": "rich",
    "url": null,
  },
}
`;

exports[`EventManagers >> GlobalTestManager >> GlobalTestManager can perform a short circuit test 1`] = `
Object {
  "type": 0,
  "value": Object {
    "author": null,
    "color": 9044739,
    "description": null,
    "fields": Array [
      Object {
        "inline": false,
        "name": "testPrefix",
        "value": "result for 5",
      },
    ],
    "footer": null,
    "image": null,
    "thumbnail": null,
    "timestamp": null,
    "title": "Trust thine instincts, however deceitful they may be",
    "type": "rich",
    "url": null,
  },
}
`;

exports[`EventManagers >> GlobalTestManager >> GlobalTestManager returns a proper value on to low of a score 1`] = `
Object {
  "type": 0,
  "value": Object {
    "author": null,
    "color": 9044739,
    "description": null,
    "fields": Array [
      Object {
        "inline": false,
        "name": "testPrefix",
        "value": "You can't seem to figure it out...",
      },
    ],
    "footer": null,
    "image": null,
    "thumbnail": null,
    "timestamp": null,
    "title": "Trust thine instincts, however deceitful they may be",
    "type": "rich",
    "url": null,
  },
}
`;
